src/main/java/frc/robot/subsystems/drive/SwerveModule.java

```diff
@@ -0,0 +1,166 @@
+                        double resetOffset,// for physically rotated modules, reset the offsets at a different angle
+                        int driveMotorCanId,
+                        int rotateMotorCanId,
+                        int rotateSensorPort)
+    {
+        super(x, y);
+
+        _rotateSetpoint       = 0;
+        _driveSpeedSetpoint   = 0;
+
+        _rotationZero         = defaultRelativeZero;
+        _resetOffset          = resetOffset;
+
+        _drivePosition        = 0;
+
+        _driveMotor           = new TalonFX(driveMotorCanId);
+        _rotateMotor          = new CANSparkMax(rotateMotorCanId, MotorType.kBrushless);
+        _rotateSensor         = new AnalogPotentiometer(rotateSensorPort, 360 / 0.92, (360 - 360 / 0.92) / 2.0); // copied from 2021
+
+        _rotateMotor.restoreFactoryDefaults(); // doesn't matter how the sparkmax has been previously set up
+
+        // the sensor is attached by a gear, so the drive motor must be inverted to counteract that
+        // rotating the module more right should result in a more positive sensor reading (heading)
+        //_rotateMotor.setInverted(false);
+
+        _rotatePID = new PIDControl();
+
+        _rotatePID.setCoefficient(Coefficient.P, 0.0, 0.01, 0.0);// 1% power for every degree of error
+        _rotatePID.setCoefficient(Coefficient.I, 0.0, 0.0, 0.0);
+        _rotatePID.setCoefficient(Coefficient.D, 0.0, 0.0, 0.0);
+
+        // rotate continuously to the nearest 180 degrees
+        _rotatePID.setInputRange(-90, 90);
+        _rotatePID.setContinuous(true);
+
+        _rotatePID.setOutputRange(-1, 1);
+
+        _rotatePID.setSetpointDeadband(2.5);// deadband recommended online by other teams
+    }
+
+    /**
+     * Drive the module (percent output)
+     * @param moduleCommand a Vector where heading represents the target rotation of the module, while the magnitude represents the percent output of the drive moter
+     * @return nothing lol
+     */
+    public void drive(Vector moduleCommand)
+    {
+        _driveSpeedSetpoint  = MathUtil.clamp(moduleCommand.getMagnitude(), -1, 1);
+        _rotateSetpoint = moduleCommand.getHeading();
+
+        double currentHeading = getHeading();
+
+        _rotatePID.setSetpoint(_rotateSetpoint, currentHeading);
+
+        double rotateSpeed = _rotatePID.calculate(currentHeading);// calling setSetpoint() then calculate() causes previous error to always be the same as the current error, meaning the Derivative coefficient will do nothing
+
+        double driveSpeed = _driveSpeedSetpoint * Math.cos(Math.toRadians(currentHeading - _rotateSetpoint));
+
+        if (_rotatePID.atSetpoint())
+        {
+            _rotateMotor.stopMotor();
+        }
+        else
+        {
+            //_rotateMotor.setVoltage(0.2 * 12);
+            _rotateMotor.setVoltage(rotateSpeed * 12);// setVoltage(speed * 12) is similar to set(speed); counteracts battery drain
```

Comment:
We should create a constant in Constants.java for this. Also, can you move the comment to the line before this?

```suggestion
            // setVoltage(speed * 12) is similar to set(speed); counteracts battery drain
            _rotateMotor.setVoltage(rotateSpeed * 12);
```

Reply:
Suggestion stands